### PR TITLE
enhance performance: don't compress memorystreams before comparing

### PR DESCRIPTION
**Pull request by micsthepick**
should make the crafting and storage UI just a bit snappier.

### CODE REVIEW

File: CrossMod/ItemCombining.cs [hunk 1]
```diff
@@ -15,7 +16,7 @@ public abstract class ItemCombining : ModType {
 
 		public abstract bool CanCombine(Item item1, Item item2);
 
-		public static bool CanCombineItems(Item item1, Item item2, bool checkPrefix = true) {
+		public static bool CanCombineItems(Item item1, Item item2, bool checkPrefix = true, ConditionalWeakTable<Item, string> savedItemTagIO = null) {
```
Comment (absoluteAquarian) [inline CrossMod/ItemCombining.cs]: This new parameter doesn't seem to be used

File: Sorting/ItemSorter.cs [hunk 2]
```diff
@@ -98,7 +100,7 @@ public static List<Item> Aggregate(AggregateContext context, CancellationToken t
 
 				List<Item> aggregate = new();
 
-				foreach (Item item in context.items.OrderBy(i => i.type).ThenBy(i => i.prefix))
+				foreach (Item item in context.items.OrderBy(i => i.type).ThenBy(i => i.prefix).ToList())
```
Comment (absoluteAquarian) [inline Sorting/ItemSorter.cs]: `.ToList()` should be avoided to prevent iterating over the collection twice

File: Utility.cs [hunk 3]
```diff
@@ -114,10 +115,19 @@ public static partial class Utility {
...
 		}
 
-		private static byte[] TagIOSave(Item item) {
-			using MemoryStream memoryStream = new();
-			TagIO.ToStream(ItemIO.Save(item), memoryStream);
-			return memoryStream.ToArray();
+		private static string TagIOSave(Item item, ConditionalWeakTable<Item, string> cache) {
+			string retVal;
+			// if present in cache, return cached value
+			if (cache != null && cache.TryGetValue(item, out retVal))
```
Comment (absoluteAquarian) [inline Utility.cs]: inline the variable declaration in the TryGetValue call

File: Utility.cs [hunk 4]
```diff
@@ -114,10 +115,19 @@ public static partial class Utility {
...
+			{
+				return retVal;
+			}
+			retVal = ItemIO.Save(item).ToString();
+			if (cache != null)
```
Comment (absoluteAquarian) [inline Utility.cs]: this `if` block could simply be `cache?.Add(item, retVal);`

File: Utility.cs [hunk 5]
```diff
@@ -96,7 +97,7 @@ public static partial class Utility {
 			}
 
 			try {
-				equal = TagIOSave(item1).SequenceEqual(TagIOSave(item2));
+				equal = TagIOSave(item1, savedItemTagIO).Equals(TagIOSave(item2, savedItemTagIO));
```
Comment (absoluteAquarian) [inline Utility.cs]: use `string == string` rather than `string.Equals(string)` for better code clarity

**Review by absoluteAquarian — CHANGES_REQUESTED** [review]

File: CrossMod/ItemCombining.cs (hunk 1 above)
Comment (absoluteAquarian) [reply CrossMod/ItemCombining.cs]: I also suggest re-adding the old definition and just have it call the new one in order to not break the API:
```cs
public static bool CanCombineItems(Item item1, Item item2, bool checkPrefix = true) => CanCombineItems(item1, item2, checkPrefix, null);
```

File: Utility.cs (hunk 3 above)
Comment (absoluteAquarian) [reply Utility.cs]: (edit: ignore this suggestion)

File: Sorting/ItemSorter.cs (hunk 2 above)
Comment (micsthepick) [reply Sorting/ItemSorter.cs]: there was a reason for me trying it, I encountered a strange Collection was modified error. Will have to investigate further, have removed it for now though

File: CrossMod/ItemCombining.cs (hunk 1 above)
Comment (micsthepick) [reply CrossMod/ItemCombining.cs]: in order to do that, I'd need to change the order so that null goes before checkPrefix, but I can do that.

Comment (micsthepick) [reply CrossMod/ItemCombining.cs]: actually, I can make both args required in that case, nevermind

Comment (micsthepick) [reply CrossMod/ItemCombining.cs]: parameter is now used and initialized in the specific circumstance

File: Utility.cs [hunk 6]
```diff
@@ -67,7 +68,7 @@ public static partial class Utility {
 			return count;
 		}
 
-		public static bool AreStrictlyEqual(Item item1, Item item2, bool checkStack = false, bool checkPrefix = true) {
+		public static bool AreStrictlyEqual(Item item1, Item item2, bool checkStack = false, bool checkPrefix = true, ConditionalWeakTable<Item, string> savedItemTagIO = null) {
```
Comment (absoluteAquarian) [inline Utility.cs]: similar suggestion to `ItemCombining.CanCombineItems()` for this Utility method

File: Utility.cs [hunk 7]
```diff
@@ -114,10 +115,15 @@ public static partial class Utility {
 			return equal;
 		}
 
-		private static byte[] TagIOSave(Item item) {
-			using MemoryStream memoryStream = new();
-			TagIO.ToStream(ItemIO.Save(item), memoryStream);
-			return memoryStream.ToArray();
+		private static string TagIOSave(Item item, ConditionalWeakTable<Item, string> cache) {
+            // if present in cache, return cached value
```
Comment (absoluteAquarian) [inline Utility.cs]: accidental spaces instead of tabs?

**Review by absoluteAquarian — CHANGES_REQUESTED** [review]

File: Utility.cs (hunk 7 above)
Comment (micsthepick) [reply Utility.cs]: gah, even with the whitespace shown, sometimes I miss it apparently!

**Review by absoluteAquarian — APPROVED** [review]

**Review by absoluteAquarian — APPROVED** [review]

**Pull request merged**